File: safari-600.7-branch/PerformanceTests/MallocBench/MallocBench/message.cpp
/*
 * Copyright (C) 2014 Apple Inc. All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 *
 * THIS SOFTWARE IS PROVIDED BY APPLE INC. ``AS IS'' AND ANY
 * EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
 * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
 * PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL APPLE INC. OR
 * CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR
 * PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY
 * OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE. 
 */

#include "CPUCount.h"
#include "message.h"
#include <dispatch/dispatch.h>
#include <stdlib.h>
#include <strings.h>

#include "mbmalloc.h"

namespace {

size_t hash(size_t hash, unsigned short a, unsigned short b)
{
    hash += a ^ b;
    return hash;
}

class Node {
    static const size_t payloadCount = 128;
public:
    void* operator new(size_t size)
    {
        return mbmalloc(size);
    }

    void operator delete(void* p, size_t size)
    {
        mbfree(p, size);
    }

    Node()
        : m_payload()
    {
    }

    size_t hash(size_t hash)
    {
        for (size_t i = 0; i < payloadCount; i += 2)
            hash = ::hash(hash, m_payload[i], m_payload[i + 1]);
        return hash;
    }

private:
    unsigned short m_payload[payloadCount];
};

class Message {
    static const size_t nodeCount = 1 * 1024;

public:
    void* operator new(size_t size)
    {
        return mbmalloc(size);
    }

    void operator delete(void* p, size_t size)
    {
        mbfree(p, size);
    }

    Message()
        : m_buffer(static_cast<Node**>(mbmalloc(nodeCount * sizeof(Node**))))
    {
        for (size_t i = 0; i < nodeCount; ++i)
            m_buffer[i] = new Node;
    }
    
    ~Message()
    {
        for (size_t i = 0; i < nodeCount; ++i)
            delete m_buffer[i];
        mbfree(m_buffer, nodeCount * sizeof(Node**));
    }

    size_t hash()
    {
        size_t hash = 0;
        for (size_t i = 0; i < nodeCount; ++i)
            hash = m_buffer[i]->hash(hash);
        return hash;
    }

private:
    Node** m_buffer;
};

} // namespace

void benchmark_message_one(bool isParallel)
{
    if (isParallel)
        abort();

    const size_t times = 2048;
    const size_t quantum = 16;

    dispatch_queue_t queue = dispatch_queue_create("message", 0);

    for (size_t i = 0; i < times; i += quantum) {
        for (size_t j = 0; j < quantum; ++j) {
            Message* message = new Message;
            dispatch_async(queue, ^{
                size_t hash = message->hash();
                if (hash)
                    abort();
                delete message;
            });
        }
        dispatch_sync(queue, ^{ });
    }

    dispatch_sync(queue, ^{ });

    dispatch_release(queue);
}

void benchmark_message_many(bool isParallel)
{
    if (isParallel)
        abort();

    const size_t times = 768;
    const size_t quantum = 16;

    const size_t queueCount = cpuCount() - 1;
    dispatch_queue_t queues[queueCount];
    for (size_t i = 0; i < queueCount; ++i)
        queues[i] = dispatch_queue_create("message", 0);

    for (size_t i = 0; i < times; i += quantum) {
        for (size_t j = 0; j < quantum; ++j) {
            for (size_t k = 0; k < queueCount; ++k) {
                Message* message = new Message;
                dispatch_async(queues[k], ^{
                    size_t hash = message->hash();
                    if (hash)
                        abort();
                    delete message;
                });
            }
        }

        for (size_t i = 0; i < queueCount; ++i)
            dispatch_sync(queues[i], ^{ });
    }

    for (size_t i = 0; i < queueCount; ++i)
        dispatch_sync(queues[i], ^{ });

    for (size_t i = 0; i < queueCount; ++i)
        dispatch_release(queues[i]);
}

File: Tools/Scripts/libraries/webkitscmpy/webkitscmpy/program/canonicalize/message.py
#!/usr/bin/env python3

# Copyright (C) 2020-2022 Apple Inc. All rights reserved.
#
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
# 1.  Redistributions of source code must retain the above copyright
#     notice, this list of conditions and the following disclaimer.
# 2.  Redistributions in binary form must reproduce the above copyright
#     notice, this list of conditions and the following disclaimer in the
#     documentation and/or other materials provided with the distribution.
#
# THIS SOFTWARE IS PROVIDED BY APPLE INC. AND ITS CONTRIBUTORS "AS IS" AND
# ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
# WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
# DISCLAIMED. IN NO EVENT SHALL APPLE INC. OR ITS CONTRIBUTORS BE LIABLE FOR
# ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
# DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
# SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
# CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY,
# OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

import os
import sys
from typing import IO

from webkitscmpy.local import Git
from webkitscmpy import Commit


def main(inputfile, identifier_template):
    REPOSITORY_PATH = os.environ.get('OLDPWD')
    GIT_COMMIT = os.environ.get('GIT_COMMIT')

    if not REPOSITORY_PATH:
        sys.stderr.write('Failed to retrieve repository path\n')
        return -1
    if not GIT_COMMIT:
        sys.stderr.write('Failed to retrieve git hash\n')
        return -1

    repository = Git(REPOSITORY_PATH)
    commit = repository.commit(hash=GIT_COMMIT)
    if not commit:
        sys.stderr.write("Failed to find '{}' in the repository".format(GIT_COMMIT))
        return -1
    if not commit.identifier or not commit.branch:
        sys.stderr.write("Failed to compute the identifier for '{}'".format(GIT_COMMIT))
        return -1

    rewrite_message(inputfile, sys.stdout, commit, identifier_template)


def rewrite_message(inputfile: IO[str], outputfile: IO[str], commit: Commit, identifier_template: str) -> None:
    lines = []
    for line in inputfile.readlines():
        lines.append(line.rstrip())

    identifier_index = len(lines)
    if identifier_index and Git.GIT_SVN_REVISION.match(lines[-1]):
        identifier_index -= 1

    # We're trying to cover cases where there is a space between link and git-svn-id:
    #     <commit message content>
    #
    #     Canonical link: ...
    #     git-svn-id: ...
    # OR
    #     <commit message content>
    #     Canonical link: ...
    #
    #     git-svn-id: ...
    if identifier_index and lines[identifier_index - 1].startswith(identifier_template.format('').split(':')[0]):
        lines[identifier_index - 1] = identifier_template.format(commit)
        identifier_index = identifier_index - 2
    else:
        for index in [2, 3]:
            if identifier_index - index > 0 and lines[identifier_index - index].startswith(identifier_template.format('').split(':')[0]):
                del lines[identifier_index - index]
                lines.insert(identifier_index - 1, identifier_template.format(commit))
                identifier_index = identifier_index - 2
                break
        else:
            lines.insert(identifier_index, identifier_template.format(commit))
            identifier_index = identifier_index - 1

    while identifier_index >= 0 and lines[identifier_index]:
        if Commit.TRAILER_RE.match(lines[identifier_index]):
            identifier_index -= 1
        else:
            lines.insert(identifier_index + 1, '')
            break

    # Turn multiple empty lines at the end of a commit into a single one
    index = len(lines) - 1
    while index and lines[index]:
        index -= 1
    index -= 1
    while index and not lines[index]:
        del lines[index]
        index -= 1

    for line in lines:
        print(line, file=outputfile)


if __name__ == '__main__':
    sys.exit(main(sys.stdin, sys.argv[1]))
